telemetry/core/src/main/scala/io/buoyant/telemetry/KafkaTelemeter.scala

```diff
@@ -0,0 +1,205 @@
+    extends SamplingTracer(tracer, sampleRate)
+
+  //Only support thrift for now.
+  class KafkaRawZipkinTracer(
+    brokerList: String,
+    numRetries: Int = 3, //this is the kafka default.
+    topic: String = "zipkin",
+    timer: Timer = DefaultTimer.twitter,
+    statsReceiver: StatsReceiver = DefaultStatsReceiver
+  ) extends RawZipkinTracer(statsReceiver, timer) {
+    log.info("Initializing kafka zipkin tracer")
+    private var producer: KafkaProducer[Array[Byte], Array[Byte]] = null
+    val configs: java.util.Map[String, AnyRef] = new java.util.HashMap[String, AnyRef]
+    configs.put(ProducerConfig.BOOTSTRAP_SERVERS_CONFIG, brokerList)
+    configs.put(ProducerConfig.KEY_SERIALIZER_CLASS_CONFIG, "org.apache.kafka.common.serialization.ByteArraySerializer")
+    configs.put(ProducerConfig.VALUE_SERIALIZER_CLASS_CONFIG, "org.apache.kafka.common.serialization.ByteArraySerializer")
+    configs.put(ProducerConfig.METADATA_FETCH_TIMEOUT_CONFIG, METADATA_FETCH_TIMEOUT_CONFIG.asInstanceOf[AnyRef])
+    configs.put(ProducerConfig.TIMEOUT_CONFIG, TIMEOUT_CONFIG.asInstanceOf[AnyRef])
+    configs.put(ProducerConfig.RETRY_BACKOFF_MS_CONFIG, RETRY_BACKOFF_MS_CONFIG.asInstanceOf[AnyRef])
+    configs.put(ProducerConfig.RECONNECT_BACKOFF_MS_CONFIG, RECONNECT_BACKOFF_MS_CONFIG.asInstanceOf[AnyRef])
+    configs.put(ProducerConfig.RETRIES_CONFIG, numRetries.asInstanceOf[AnyRef])
+    producer = new KafkaProducer[Array[Byte], Array[Byte]](configs)
+
+    //retries configured from the input.
+    private[this] val scopedReceiver = statsReceiver.scope("log_span")
+    private[this] val okCounter = scopedReceiver.counter("ok")
+    private[this] val errorReceiver = scopedReceiver.scope("error")
+
+    private[this] def serializeBytes(span: Span): Array[Byte] = {
+      log.debug("Serializing span data to bytes")
+      val tSerializer = new TSerializer()
+      tSerializer.serialize(span.toThrift)
+    }
+
+    /**
+     * Log the span data via kafka.
+     */
+    def sendSpans(spans: Seq[Span]): Future[Unit] = {
+      //delegate batching to downstream for simplicity reasons rather than doing buffer pools here.
+      //The one complication it adds is it is hard to do retries. Breaking it into the model if any one stat emission
+      //fails the entire batch is considered to be failed. The batch will only fail if the retries for all the different
+      //stats emitted as part of the batch are exhausted independently of each other
+      // without being able to send it out successfully.
+      // Will try to do early detection. Currently this will enable both
+      //kafka level and higher level retries if the framework supports it.
+      var ex: Throwable = null
+      Future({
+        log.debug("Sending kafka %i trace events", spans.size)
+        val statEmissionFailures: AtomicBoolean = new AtomicBoolean(false)
+        val semaphore: Semaphore = new Semaphore((spans.size - 1) * -1)
```

Comment:
I think you skipped this part of the comments  "  // Will try to do early detection." 
Currently if even a single event in the batch fails then we consider it a failure without waiting for the others to succeed.